pkg/indexerpoc/source.go

```diff
@@ -0,0 +1,216 @@
+package indexerpoc
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"go.uber.org/zap"
+)
+
+// GethSource is an implementation of the Source interface using go-ethereum.
+type GethSource struct {
+	client    *ethclient.Client
+	rpcClient *rpc.Client
+	network   *Network
+	log       *zap.Logger
+
+	// Cache for latest block
+	cacheMu      sync.Mutex
+	latestNumber uint64
+	latestHash   common.Hash
+	cacheHits    int
+	maxCacheHits int
+}
+
+// NewGethSource creates a new Ethereum client source.
+func NewGethSource(
+	network *Network,
+	log *zap.Logger,
+) (*GethSource, error) {
+	if network == nil {
+		return nil, fmt.Errorf("network configuration cannot be nil")
+	}
+
+	// Create a network-specific named logger
+	networkLogger := log.Named("network").With(zap.String("network", network.Name))
+
+	rpcClient, err := rpc.Dial(network.RpcURL)
+	if err != nil {
+		return nil, fmt.Errorf("connecting to %s node: %w", network.Name, err)
+	}
+
+	client := ethclient.NewClient(rpcClient)
+
+	// Verify chain ID matches the expected one
+	chainID, err := client.ChainID(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("getting chain ID from %s: %w", network.Name, err)
+	}
+
+	if chainID.Int64() != network.ChainID {
+		return nil, fmt.Errorf("chain ID mismatch for %s: expected %d, got %d",
+			network.Name, network.ChainID, chainID.Int64())
+	}
+
+	source := &GethSource{
+		client:       client,
+		rpcClient:    rpcClient,
+		network:      network,
+		log:          networkLogger,
+		maxCacheHits: 20,
+	}
+
+	// Start a background goroutine to keep track of the latest block
+	go source.pollLatestBlock(context.Background())
+
+	return source, nil
+}
+
+// GetNetworkName returns the name of the network
+func (s *GethSource) GetNetworkName() string {
+	return s.network.Name
+}
+
+// GetChainID returns the chain ID of the network
+func (s *GethSource) GetChainID() int64 {
+	return s.network.ChainID
+}
+
+// pollLatestBlock periodically polls for the latest block
+func (s *GethSource) pollLatestBlock(ctx context.Context) {
```

Comment:
do we still need a polling indexer? we should be using the Alchemy WSS interface